flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java

```diff
@@ -0,0 +1,228 @@
+    this.errorCounter =
+        TableMaintenanceMetrics.groupFor(getRuntimeContext(), tableName, taskName, taskIndex)
+            .counter(TableMaintenanceMetrics.ERROR_COUNTER);
+
+    this.rewriter =
+        new SizeBasedDataRewriter(tableLoader.loadTable()) {
+          @Override
+          public Set<DataFile> rewrite(List<FileScanTask> group) {
+            // We use the rewriter only for bin-packing the file groups to compact
+            throw new UnsupportedOperationException("Should not be called");
+          }
+        };
+
+    rewriter.init(rewriterOptions);
+    this.planner = new RewriteFileGroupPlanner(rewriter, RewriteJobOrder.NONE);
+  }
+
+  @Override
+  public void processElement(Trigger value, Context ctx, Collector<PlannedGroup> out)
+      throws Exception {
+    LOG.debug(
+        "Creating rewrite plan for table {} with {}[{}] at {}",
+        tableName,
+        taskName,
+        taskIndex,
+        ctx.timestamp());
+    try {
+      SerializableTable table =
+          (SerializableTable) SerializableTable.copyOf(tableLoader.loadTable());
+      if (table.currentSnapshot() == null) {
+        LOG.info(
+            "Nothing to plan for in an empty table {} with {}[{}] at {}",
+            tableName,
+            taskName,
+            taskIndex,
+            ctx.timestamp());
+        return;
+      }
+
+      RewritePlanResult plan =
+          planner.plan(table, Expressions.alwaysTrue(), table.currentSnapshot().snapshotId());
+
+      long rewriteBytes = 0;
+      List<RewriteFileGroup> groups = plan.fileGroups().collect(Collectors.toList());
+      ListIterator<RewriteFileGroup> iter = groups.listIterator();
+      while (iter.hasNext()) {
+        RewriteFileGroup group = iter.next();
+        if (rewriteBytes + group.sizeInBytes() > maxRewriteBytes) {
+          // Keep going, maybe some other group might fit in
+          LOG.info(
+              "Skipping group {} as max rewrite size reached for table {} with {}[{}] at {}",
+              group,
+              tableName,
+              taskName,
+              taskIndex,
+              ctx.timestamp());
+          iter.remove();
+        } else {
+          rewriteBytes += group.sizeInBytes();
+        }
+      }
+
+      int groupsPerCommit =
+          IntMath.divide(
+              plan.context().totalGroupCount(), partialProgressMaxCommits, RoundingMode.CEILING);
+
+      LOG.info(
+          "Rewrite plan created {} for table {} with {}[{}] at {}",
+          groups,
+          tableName,
+          taskName,
+          taskIndex,
+          ctx.timestamp());
+
+      for (RewriteFileGroup group : groups) {
+        LOG.debug(
+            "Emitting {} with for table {} with {}[{}] at {}",
+            group,
+            tableName,
+            taskName,
+            taskIndex,
+            ctx.timestamp());
+        out.collect(
+            new PlannedGroup(
+                table, groupsPerCommit, rewriter.splitSize(group.sizeInBytes()), group));
```

Comment:
why do we need to calculate split size here via `rewriter.splitSize()`? target split size is needed for scan planning. this is already after planning.

Reply:
This is for calculating the target file size for the rewriter. Might change after the core refactor PR. We will have a place to store it in the plan